ext/opentelemetry-ext-flask/src/opentelemetry/ext/flask/__init__.py

```diff
@@ -1,97 +1,124 @@
+
+            def _start_response(status, response_headers, *args, **kwargs):
+                span = flask.request.environ.get(_ENVIRON_SPAN_KEY)
+                if span:
+                    otel_wsgi.add_response_attributes(
+                        span, status, response_headers
+                    )
+                else:
+                    logger.warning(
+                        "Flask environ's OpenTelemetry span "
+                        "missing at _start_response(%s)",
+                        status,
+                    )
+
+                return start_response(
+                    status, response_headers, *args, **kwargs
+                )
+
+            return wsgi(environ, _start_response)
+
+        self.wsgi_app = wrapped_app
+
+        @self.before_request
+        def _before_flask_request():
+            environ = flask.request.environ
+            span_name = (
+                flask.request.endpoint
+                or otel_wsgi.get_default_span_name(environ)
+            )
+            parent_span = propagators.extract(
+                otel_wsgi.get_header_from_environ, environ
+            )
+
+            tracer = trace.tracer_source().get_tracer(__name__, __version__)
+
+            attributes = otel_wsgi.collect_request_attributes(environ)
+            if flask.request.url_rule:
+                # For 404 that result from no route found, etc, we
+                # don't have a url_rule.
+                attributes["http.route"] = flask.request.url_rule.rule
+            span = tracer.start_span(
+                span_name,
+                parent_span,
+                kind=trace.SpanKind.SERVER,
+                attributes=attributes,
+                start_time=environ.get(_ENVIRON_STARTTIME_KEY),
+            )
+            activation = tracer.use_span(span, end_on_exit=True)
+            activation.__enter__()
+            environ[_ENVIRON_ACTIVATION_KEY] = activation
+            environ[_ENVIRON_SPAN_KEY] = span
+
+        @self.teardown_request
+        def _teardown_flask_request(exc):
+            activation = flask.request.environ.get(_ENVIRON_ACTIVATION_KEY)
+            if not activation:
+                logger.warning(
+                    "Flask environ's OpenTelemetry activation missing"
+                    "at _teardown_flask_request(%s)",
+                    exc,
+                )
+                return
+
+            if exc is None:
+                activation.__exit__(None, None, None)
+            else:
+                activation.__exit__(
+                    type(exc), exc, getattr(exc, "__traceback__", None)
+                )
+
 
-    You must not call this function multiple times on the same Flask object.
+@BasePatcher.protect
+class FlaskPatcher(BasePatcher):
+    """A patcher for flask.Flask
+
+    See `BasePatcher`
     """
 
-    wsgi = flask.wsgi_app
+    def __init__(self):
+        self._patched = False
+        self._original_flask = flask.Flask
```

Comment:
I would capture that in `patch`. Otherwise you might undo more than you want in `unpatch`.

Reply:
Makes sense